### PR TITLE
[MU4] Fix #7818 Added a limit to canvas scroll

### LINKED ISSUE #7818: [MU4 Issue] Shouldn't be able to scroll past edge of score
**Describe the bug**
Score is scrolling past edge

**To Reproduce**
1. Create a score
2. Try to scroll past edge of score

**Expected behavior**
Shouldn't be able to scroll past edge of score

**Screenshots**

**Desktop (please complete the following information):**
MacOS
Wndows 10

**Additional context**

### PR DESCRIPTION
**Pull request by Shivansh-Bakshi**
Resolves: #7818 

Limits the canvas scroll under Limit Scroll on and off conditions.. Tested with zoom levels and restore down on windows.

- [x] I signed [CLA](https://musescore.org/en/cla)
- [x] I made sure the code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [x] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [ ] I created the test (mtest, vtest, script test) to verify the changes I made

### CODE REVIEW

**Comment by vshalkevich** [comment]
Works fine for me!

File: src/notation/view/notationpaintview.cpp [hunk 1]
```diff
@@ -388,20 +388,29 @@ std::pair<int, int> NotationPaintView::constraintCanvas(int dx, int dy) const
...
 
     QPoint canvasCenter = this->canvasCenter();
-    if (contentRect.width() <= viewport.width()) {
+
+    bool isScrollLimited = configuration()->isLimitCanvasScrollArea();
+
+    int offsetX = viewport.width() - 100;
+    int offsetY = viewport.height() - 100;
+    if (isScrollLimited) {
+        offsetX = offsetY = 100;
```
Comment (RomanPudashkin) [inline src/notation/view/notationpaintview.cpp:397]: I'm afraid this fix won't work on hidpi screens. Anyway, it is better don't use any magic numbers here. You should use adaptive values dependent on the current screen width/height/scaling

Comment (Shivansh-Bakshi) [reply src/notation/view/notationpaintview.cpp:397]: Right. I'll change the 100 to a relative value. I don't understand why it won't work on hidpi screens though. Could you please explain why?

Comment (RomanPudashkin) [reply src/notation/view/notationpaintview.cpp]: This is just my assumption, I don't have any proofs :) But magic numbers always have such kinds of problems (especially in the score view where everything depends on the current screen configuration). So better always use relative values

File: src/notation/view/notationpaintview.cpp [hunk 2]
```diff
@@ -388,20 +388,29 @@ std::pair<int, int> NotationPaintView::constraintCanvas(int dx, int dy) const
     QRectF viewport = this->viewport();
 
     QPoint canvasCenter = this->canvasCenter();
-    if (contentRect.width() <= viewport.width()) {
+
+    bool isScrollLimited = configuration()->isLimitCanvasScrollArea();
+
+    int offsetX = viewport.width() * 0.98;
+    int offsetY = viewport.height() * 0.98;
```
Comment (RomanPudashkin) [inline src/notation/view/notationpaintview.cpp:395]: Better to add some named constants for those values (static constexpr at the top of this file)

Comment (Shivansh-Bakshi) [reply src/notation/view/notationpaintview.cpp]: Done. I should've done that first. Beginner mistake 😅

**Comment by Eism** [comment]
the viewable area is too small(Windows)

<details>
<summary>Screenshot</summary>

image

</details>

**Comment by Shivansh-Bakshi** [comment]
> the viewable area is too small(Windows)
> 
> Screenshot
> image

Right now it allows you to scroll till 98% of the background is visible. Is there an amount you would recommend that would look nice?

**Comment by vshalkevich** [comment]
@shoogle need your advice about this fix. Is it look good or need to left more score view ?

**Comment by shoogle** [comment]
@vshalkevich, see https://github.com/musescore/MuseScore/issues/7818#issuecomment-814889174. I'd say we can probably go up to 75% background, 25% score. I don't see a reason to allow much more than that.

**Comment by Shivansh-Bakshi** [comment]
75%: 
Screenshot (118)

Screenshot (119)

More than enough for having 2-3 windows open.


50% seems good too:
Screenshot (120)

Screenshot (121)

**Comment by shoogle** [comment]
@Shivansh-Bakshi, let's go with the 75% option. There's bound to be some edge case with multiple/ultrawide monitors that we haven't thought of, so let's allow people a bit of extra room for the "no limit" option.

**Comment by Shivansh-Bakshi** [comment]
Done. Set it 75%

**Comment by vshalkevich** [comment]
For some reason I can`t run "musescore4.exe" on Windows. 
Other ".exe" builds opens...

**Comment by Shivansh-Bakshi** [comment]
@vshalkevich You mean after the changes from this PR or in general? Either ways, it seems to open just fine for me. Everytime I  build (through QtCreator or through cmake directly, release or debug), I get an executable named MuseScore4.exe. How are the other executables different?

**Comment by shoogle** [comment]
@vshalkevich

> For some reason I can`t run "musescore4.exe" on Windows.

The nightlies haven been broken on Windows and Linux since the 6th of April so I don't think it is this PR's fault.

> Other ".exe" builds opens...

When were they compiled? If you hover the cursor over the build time on the Artifact page then it shows the date it was compiled. If it was before the 6th of April then it was before the nightlies broke.

image

**Comment by vshalkevich** [comment]
Need to get the latest fixes for Windows and Linux platforms.
This PR still unable to launch for me

**Comment by Shivansh-Bakshi** [comment]
@vshalkevich Should I rebase to the current master, make these changes and push it to this PR's branch (after squishing the commits)?

**Comment by RomanPudashkin** [comment]
> @vshalkevich Should I rebase to the current master, make these changes and push it to this PR's branch (after squishing the commits)?

Yes, you should just rebase to the current master

**Comment by vshalkevich** [comment]
Limitation scroll works  but Score View is black 
2021-04-14 17_42_54-NVIDIA GeForce Overlay

**Comment by Shivansh-Bakshi** [comment]
That's not happening for me. The changes from this PR only adds a constraint to the canvas. It does not change anything on the score directly.
Constraint

**Comment by vshalkevich** [comment]
@shoogle 
Could you check is score view black for you in this PR or normal ?

**Comment by vshalkevich** [comment]
Works fine in all directions on MacOS

**Comment by vshalkevich** [comment]
Works fine for me in all platforms. Black screen gone, this issue was not from this PR

**Review by RomanPudashkin — APPROVED** [review]

**Pull request merged**